src/base/Algorithms.h

```diff
@@ -24,6 +24,21 @@ auto transform(const InputVector &input, Converter convert)
     return output;
 }
 
+template<typename OutputVector, typename InputVector, typename Converter>
+auto filtered(const InputVector &input, Converter convert)
+{
+    OutputVector output;
+    if constexpr (std::ranges::sized_range<InputVector>) {
+        output.reserve(input.size());
+    }
+    for (const auto &value : input) {
+        if (const std::optional<std::decay_t<decltype(value)>> result = std::invoke(convert, value)) {
+            output.push_back(*result);
+        }
+    }
+    return output;
+}
```

Comment:
Since you're using the capacity of the old vector (without filtering), I'd suggest to add a call to `output.shrink_to_fit()` in case many elements have been removed.